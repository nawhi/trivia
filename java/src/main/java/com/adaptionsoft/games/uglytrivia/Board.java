package com.adaptionsoft.games.uglytrivia;

public class Board {
    int[] places = new int[6];

    public Board() {
    }

    void movePlayer(int roll, int currentPlayerIndex) {
        places[currentPlayerIndex] = places[currentPlayerIndex] + roll;
        if (places[currentPlayerIndex] > 11)
            places[currentPlayerIndex] = places[currentPlayerIndex] - 12;
    }

    int getPlayerPlace(int currentPlayerIndex) {
        return places[currentPlayerIndex];
    }

    void initialisePlayerAt(int playerIndex) {
        places[playerIndex] = 0;
    }
}