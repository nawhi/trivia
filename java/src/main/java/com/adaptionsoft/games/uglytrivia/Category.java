package com.adaptionsoft.games.uglytrivia;

import java.util.LinkedList;

public class Category {
    private LinkedList<String> questions;
    private String name;
    private int questionIndex;

    public Category(String name) {
        questionIndex = 0;
        this.name = name;
        questions = new LinkedList<>();
    }

    public String nextQuestion() {
        return questions.removeFirst();
    }

    public void addQuestion() {
        questions.add(name + " Question " + questionIndex++);
    }

    boolean hasName(String name) {
        return this.name.equals(name);
    }
}