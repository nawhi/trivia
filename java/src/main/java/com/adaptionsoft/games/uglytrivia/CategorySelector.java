package com.adaptionsoft.games.uglytrivia;

public class CategorySelector {
    public String invoke(int place) {
        if (place > 11)
            throw new IllegalArgumentException("place should be less than 12");

        switch(place % 4)
        {
            case 0:  return "Pop";
            case 1:  return "Science";
            case 2:  return "Sports";
            default: return "Rock";
        }
    }
}
