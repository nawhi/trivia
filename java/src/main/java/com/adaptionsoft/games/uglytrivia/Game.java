package com.adaptionsoft.games.uglytrivia;

import java.util.ArrayList;

public class Game {
	private final CategorySelector categorySelector = new CategorySelector();
	Players playersAggregate = new Players();
    int[] places = new int[6];
    int[] purses  = new int[6];
    boolean[] inPenaltyBox  = new boolean[6];

    int currentPlayer = 0;
    boolean isGettingOutOfPenaltyBox;
	private final QuestionPool questionPool;

	public  Game(){
		questionPool = new QuestionPool("Pop", "Science", "Sports", "Rock");
	}

	public boolean add(String playerName) {
	    playersAggregate.add(new Player(playerName));
		places[playersAggregate.numberOfPlayers()] = 0;
		purses[playersAggregate.numberOfPlayers()] = 0;
		inPenaltyBox[playersAggregate.numberOfPlayers()] = false;
	    
	    System.out.println(playerName + " was added");
	    System.out.println("They are player number " + playersAggregate.numberOfPlayers());
		return true;
	}

	public void roll(int roll) {
		System.out.println(playersAggregate.getNameByIndex(currentPlayer) + " is the current player");
		System.out.println("They have rolled a " + roll);
		
		if (inPenaltyBox[currentPlayer]) {
			if (roll % 2 != 0) {
				isGettingOutOfPenaltyBox = true;
				
				System.out.println(playersAggregate.getNameByIndex(currentPlayer) + " is getting out of the penalty box");
				places[currentPlayer] = places[currentPlayer] + roll;
				if (places[currentPlayer] > 11) places[currentPlayer] = places[currentPlayer] - 12;
				
				System.out.println(playersAggregate.getNameByIndex(currentPlayer)
						+ "'s new location is " 
						+ places[currentPlayer]);
				System.out.println("The category is " + categorySelector.categoryFor(places[currentPlayer]));
				askQuestion();
			} else {
				System.out.println(playersAggregate.getNameByIndex(currentPlayer) + " is not getting out of the penalty box");
				isGettingOutOfPenaltyBox = false;
				}
			
		} else {
		
			places[currentPlayer] = places[currentPlayer] + roll;
			if (places[currentPlayer] > 11) places[currentPlayer] = places[currentPlayer] - 12;
			
			System.out.println(playersAggregate.getNameByIndex(currentPlayer)
					+ "'s new location is " 
					+ places[currentPlayer]);
			System.out.println("The category is " + categorySelector.categoryFor(places[currentPlayer]));
			askQuestion();
		}
		
	}

	private void askQuestion() {
		String category = categorySelector.categoryFor(places[currentPlayer]);
		System.out.println(questionPool.questionFor(category));
	}


    public boolean wasCorrectlyAnswered() {
		if (inPenaltyBox[currentPlayer]){
			if (isGettingOutOfPenaltyBox) {
				System.out.println("Answer was correct!!!!");
				purses[currentPlayer]++;
				System.out.println(playersAggregate.getNameByIndex(currentPlayer)
						+ " now has "
						+ purses[currentPlayer]
						+ " Gold Coins.");
				
				boolean winner = didPlayerWin();
				currentPlayer++;
				if (currentPlayer == playersAggregate.numberOfPlayers()) currentPlayer = 0;
				
				return winner;
			} else {
				currentPlayer++;
				if (currentPlayer == playersAggregate.numberOfPlayers()) currentPlayer = 0;
				return true;
			}
			
			
			
		} else {
		
			System.out.println("Answer was corrent!!!!");
			purses[currentPlayer]++;
			System.out.println(playersAggregate.getNameByIndex(currentPlayer)
					+ " now has "
					+ purses[currentPlayer]
					+ " Gold Coins.");
			
			boolean winner = didPlayerWin();
			currentPlayer++;
			if (currentPlayer == playersAggregate.numberOfPlayers()) currentPlayer = 0;
			
			return winner;
		}
	}
	
	public boolean wrongAnswer(){
		System.out.println("Question was incorrectly answered");
		System.out.println(playersAggregate.getNameByIndex(currentPlayer)+ " was sent to the penalty box");
		inPenaltyBox[currentPlayer] = true;
		
		currentPlayer++;
		if (currentPlayer == playersAggregate.numberOfPlayers()) currentPlayer = 0;
		return true;
	}


	private boolean didPlayerWin() {
		return !(purses[currentPlayer] == 6);
	}

}
