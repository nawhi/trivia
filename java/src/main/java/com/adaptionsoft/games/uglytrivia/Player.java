package com.adaptionsoft.games.uglytrivia;

import java.util.Objects;

public class Player {
    private String name;
    private int purse;

    public String getName() {
        return name;
    }

    public Player(String name) {
        this.name = name;
    }

    public void incrementPurse() {
        purse++;
    }

    public int getPurse() {
        return purse;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Player player = (Player) o;
        return Objects.equals(name, player.name);
    }

    @Override
    public int hashCode() {
        return Objects.hash(name);
    }

    @Override
    public String toString() {
        return "Player{" +
                "name='" + name + '\'' +
                '}';
    }
}
