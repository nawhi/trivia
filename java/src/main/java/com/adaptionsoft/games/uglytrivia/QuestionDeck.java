package com.adaptionsoft.games.uglytrivia;

import java.util.LinkedList;

public class QuestionDeck {
    private LinkedList<String> questions;
    private String category;
    private int questionIndex;

    public QuestionDeck(String category) {
        questionIndex = 0;
        this.category = category;
        questions = new LinkedList<>();
    }

    public String next() {
        return questions.removeFirst();
    }

    public void add() {
        questions.add(category + " Question " + questionIndex++);
    }
}