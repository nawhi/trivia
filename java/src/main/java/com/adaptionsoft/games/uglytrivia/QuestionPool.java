package com.adaptionsoft.games.uglytrivia;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class QuestionPool {
    private List<Category> categories;

    public QuestionPool(String... categories) {
        this.categories = Arrays.stream(categories)
                .map(this::createNewQuestionsForCategory)
                .collect(Collectors.toList());
    }

    private Category createNewQuestionsForCategory(String categoryName) {
        Category questionsForCategory = new Category(categoryName);
        IntStream.range(0, 50).forEach(i -> questionsForCategory.addQuestion());
        return questionsForCategory;
    }

    public String questionFor(String category) {
        return categories.stream()
                .filter(questionsForCategory -> questionsForCategory.hasName(category))
                .findFirst()
                .orElseThrow(IllegalArgumentException::new)
                .nextQuestion();
    }
}