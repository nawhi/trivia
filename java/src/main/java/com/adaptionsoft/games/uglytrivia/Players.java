package com.adaptionsoft.games.uglytrivia;

import com.adaptionsoft.games.trivia.Player;

import java.util.ArrayList;
import java.util.List;

public class Players {
    private List<Player> listOfPlayers;

    public Players() {
        listOfPlayers = new ArrayList<>();
    }

    public int numberOfPlayers() {
        return listOfPlayers.size();
    }

    public void add(Player player) {
        listOfPlayers.add(player);
    }

    public Player getByIndex(int index) {
        return listOfPlayers.get(index);
    }
}
