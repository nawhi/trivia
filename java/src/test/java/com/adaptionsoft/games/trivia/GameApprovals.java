package com.adaptionsoft.games.trivia;

import com.adaptionsoft.games.uglytrivia.Game;
import org.approvaltests.Approvals;
import org.junit.Test;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;

public class GameApprovals {
    @Test
    public void add_a_player() {
        String output = catchSystemOutFor(() -> new Game().add("Test"));
        Approvals.verify(output);
    }

    @Test
    public void add_two_players() {
        String output = catchSystemOutFor(() -> {
            Game game = new Game();
            game.add("Test1");
            game.add("Test2");
        });
        Approvals.verify(output);
    }

    @Test
    public void a_player_rolls_a_one() {
        String output = catchSystemOutFor(() -> {
            Game game = new Game();

            game.add("Test");

            game.roll(1);
        });

        Approvals.verify(output);
    }

    @Test
    public void a_player_rolls_a_six_and_answers_correctly() {
        String output = catchSystemOutFor(() -> {
            Game game = new Game();

            game.add("Test");

            game.roll(6);
            game.wasCorrectlyAnswered();
        });

        Approvals.verify(output);
    }

    @Test
    public void a_player_rolls_a_four_and_answers_incorrectly() {
        String output = catchSystemOutFor(() -> {
            Game game = new Game();

            game.add("Test");

            game.roll(4);
            game.wrongAnswer();
        });

        Approvals.verify(output);
    }

    @Test
    public void a_player_in_the_penalty_box_rolls_a_one() {
        String output = catchSystemOutFor(() -> {
            Game game = aGameWithPlayerInPenaltyBox();
            game.roll(1);
        });

        Approvals.verify(output);
    }

    @Test
    public void a_player_in_the_penalty_box_rolls_a_two() {
        String output = catchSystemOutFor(() -> {
            Game game = aGameWithPlayerInPenaltyBox();
            game.roll(2);
        });

        Approvals.verify(output);
    }

    private Game aGameWithPlayerInPenaltyBox() {
        Game game = new Game();
        game.add("Test");
        game.roll(4);
        game.wrongAnswer();
        return game;
    }

    private static String catchSystemOutFor(Runnable action) {
        ByteArrayOutputStream baos = new ByteArrayOutputStream();
        PrintStream outputStream = new PrintStream(baos);

        PrintStream originalOut = System.out;
        System.setOut(outputStream);

        try {
            action.run();
        } catch (Exception ex) {
            ex.printStackTrace();
        } finally {
            System.out.flush();
            System.setOut(originalOut);
        }

        return baos.toString();
    }
}
