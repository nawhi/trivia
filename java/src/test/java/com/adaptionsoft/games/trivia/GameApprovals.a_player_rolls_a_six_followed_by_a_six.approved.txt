Test was added
They are player number 1
Test is the current player
They have rolled a 6
Test's new location is 6
The category is Sports
Sports Question 0
Test is the current player
They have rolled a 6
Test's new location is 0
The category is Pop
Pop Question 0
